crates/oxc_linter/src/rules/typescript/consistent_indexed_object_style.rs

```diff
@@ -0,0 +1,553 @@
+                                }
+                            }
+                        }
+                    }
+                }
+                TSType::TSTypeReference(decl) => {
+                    if let TSTypeName::IdentifierReference(iden) = &decl.type_name {
+                        if iden.name == "Record"
+                            && decl.type_parameters.is_some()
+                            && self.is_index_signature
+                        {
+                            ctx.diagnostic(ConsistentIndexedObjectStyleDiagnostic(
+                                "index signature",
+                                "record",
+                                decl.span,
+                            ));
+                        }
+                    }
+
+                    for param in &decl.type_parameters {
+                        for p in &param.params {
+                            if !self.is_index_signature {
+                                if let TSType::TSTypeLiteral(lit) = p {
+                                    if lit.members.len() == 1 {
+                                        if let TSSignature::TSIndexSignature(idx) = &lit.members[0]
+                                        {
+                                            ctx.diagnostic(ConsistentIndexedObjectStyleDiagnostic(
+                                                "record",
+                                                "index signature",
+                                                idx.span,
+                                            ));
+                                        }
+                                    }
+                                }
+                            }
+
+                            if self.is_index_signature {
+                                if let TSType::TSTypeReference(r) = p {
+                                    if let TSTypeName::IdentifierReference(iden) = &r.type_name {
+                                        if iden.name == "Record" {
+                                            ctx.diagnostic(ConsistentIndexedObjectStyleDiagnostic(
+                                                "index signature",
+                                                "record",
+                                                r.span,
+                                            ));
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+                _ => (),
+            },
+            AstKind::Program(prog) => {
+                for body in &prog.body {
+                    if let Statement::FunctionDeclaration(func) = body {
+                        if let Some(return_type) = &func.return_type {
+                            if !self.is_index_signature {
+                                if let TSType::TSTypeLiteral(lit) = &return_type.type_annotation {
```

Comment:
Perhaps we don't need to revisit `TSType::TSTypeLiteral`? Since all `TSType::TSTypeLiteral` has been visited.

```rust
fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
      match node.kind() {
          AstKind::TSTypeLiteral(lit) => { } // here will visit all TSTypeLiteral in our file
      }
}
```